<!-- -->
<launch>  
  <arg name="prefix" default="racecar"/>
  <arg name="fixed_frame_id" default="odom"/> <!-- set "map" when a map is used! -->
  <arg name="debug" default="false"/>
   
  <group ns="$(arg prefix)">

    <!-- uncompress raspicam image -->
    <node pkg="image_transport" type="republish" name="republish_raspicam" args="compressed in:=raspicam_node/image raw out:=raspicam_node/image"/>
    <!-- convert scan to a point cloud -->
    <node pkg="racecar_behaviors" type="laserscan_to_pointcloud.py" name="laserscan_to_pointcloud">
       <remap from="converted_pc" to="scan_cloud"/>
       <remap from="/scan" to="scan"/>
    </node>
    <!-- register scan to camera -->
    <node pkg="rtabmap_ros" type="pointcloud_to_depthimage" name="pointcloud_to_depthimage">
       <param name="fixed_frame_id"  value="$(arg prefix)/odom"/>
       <param name="fill_holes_size" value="2"/>
       <remap from="camera_info" to="raspicam_node/camera_info"/>
       <remap from="cloud"     to="scan_cloud"/>
       <remap from="image"     to="raspicam_node/depth_registered"/>
       <remap from="image_raw" to="raspicam_node/depth_registered_raw"/>
    </node>
    <!-- blob_detector -->
    <node pkg="racecar_behaviors" type="blob_detector.py" name="blob_detector" output="screen">
       <remap from="image"       to="raspicam_node/image"/>
       <remap from="camera_info" to="raspicam_node/camera_info"/>
       <remap from="depth"       to="raspicam_node/depth_registered"/>
       <param name="map_frame_id"    value="$(arg prefix)/$(arg fixed_frame_id)"/>
       <param name="frame_id"        value="$(arg prefix)/base_footprint"/>
       <param name="object_frame_id" value="$(arg prefix)/object"/>
       <param name="color_hue"       value="10"/>
    </node>
        
    <!-- used to debug camera vs scan registration -->
    <group if="$(arg debug)">
       <node pkg="nodelet" type="nodelet" name="point_cloud_xyzrgb" args="standalone rtabmap_ros/point_cloud_xyzrgb">
          <param name="approx_sync"  value="false"/>
          <remap from="cloud"           to="raspicam_node/registered_cloud"/>
          <remap from="depth/image"     to="raspicam_node/depth_registered_raw"/>
          <remap from="rgb/image"       to="raspicam_node/image_rect_color"/>
          <remap from="rgb/camera_info" to="raspicam_node/camera_info"/>
       </node>
       <node pkg="image_view" type="image_view" name="image_view_detections">
          <remap from="image"       to="/racecar/image_detections"/>
          <param name="autosize"    value="true"/>
          <param name="window_name" value="objects"/>
       </node>
       <node pkg="image_view" type="image_view" name="image_view_depth">
          <remap from="image"       to="raspicam_node/depth_registered"/>
          <param name="autosize"    value="true"/>
          <param name="window_name" value="lidar"/>
       </node>
       <node pkg="image_view" type="image_view" name="image_view_image">
          <remap from="image"       to="raspicam_node/image"/>
          <param name="autosize"    value="true"/>
          <param name="window_name" value="raw"/>
       </node>
       <group ns="raspicam_node">
          <node pkg="image_proc" type="image_proc" name="image_proc_rectify">
             <remap from="image_raw" to="image"/>
          </node>
       </group>
       <node pkg="image_view" type="image_view" name="image_view_rectified">
          <remap from="image"       to="raspicam_node/image_rect_color"/>
          <param name="autosize"    value="true"/>
          <param name="window_name" value="rectified"/>
       </node>
       <node pkg="rqt_reconfigure" type="rqt_reconfigure" name="rqt_reconfigure"/>
       <node name="rviz" pkg="rviz" type="rviz" args="-d $(find racecar_bringup)/config/bringup.rviz"/>
    </group>
        
  </group>

</launch>
